Editorial: Define _response map_ (#1143)
graphql/graphql-spec: spec/Section 7 -- Response.md

## Response Format

:: A GraphQL _response_ is either a _response map_ or a _response stream_.

### Response Map

> :: A GraphQL request returns a _response map_ when the GraphQL operation is a
> query or mutation. A _response map_ must be a map.

Reviewer: A response map can be used in many situations. We should try and make the description more accurate. I don't particularly like this edit, but it does seem more accurate.

```suggestion
:: A _response map_ is a map containing data and/or errors returned from GraphQL
execution.

Note: A _response map_ is typically the result type that you would see for a
query or mutation operation, however you may also see it in response to a
subscription operation should a request error occur.
```

the request completed without raising any errors, this entry must not be
present.

If the request included execution, the _response map_ must contain an entry with
key `data`. The value of this entry is described in the "Data" section. If the
request failed before execution, due to a syntax error, missing information, or
validation error, this entry must not be present.

The _response map_ may also contain an entry with key `extensions`. This entry,
if set, must have a map as its value. This entry is reserved for implementers to
extend the protocol however they see fit, and hence there are no additional
restrictions on its contents.

To ensure future changes to the protocol do not break existing services and
clients, the top level _response map_ must not contain any entries other than
the three described above.

Note: When `errors` is present in the _response map_, it may be helpful for it
to appear first when serialized to make it more clear when errors are present in
a _response map_ during debugging.

### Response Stream

:: A GraphQL request returns a _response stream_ when the GraphQL operation is a

> subscription. A _response stream_ must be a stream of _response map_.

Reviewer: This isn't actually true, a subscription can yield a _response map_ if a request error occurs. This should be reflected in the response map definition too.

> subscription. A _response stream_ must be a stream of _response map_.

Reviewer: ```suggestion
subscription and the subscription is successfully established (no request error
was raised). A _response stream_ must be a stream of _response map_.
```

if set, must have a map as its value. This entry is reserved for implementers to
extend the protocol however they see fit, and hence there are no additional
restrictions on its contents.

To ensure future changes to the protocol do not break existing services and
clients, the top level _response map_ must not contain any entries other than
the three described above.

Note: When `errors` is present in the _response map_, it may be helpful for it
to appear first when serialized to make it more clear when errors are present in
a response map during debugging.

### Response Stream

:: A GraphQL request returns a _response stream_ when the GraphQL operation is a
subscription and a _request error_ is not raised. A response stream must be a
stream of _response map_.

### Data

The `data` entry in the _response map_ will be the result of the execution of
the requested operation. If the operation was a query, this output will be an
object of the query root operation type; if the operation was a mutation, this
output will be an object of the mutation root operation type.

If an error was raised before execution begins, the `data` entry should not be
present in the _response map_.

> If an error was raised during the execution that prevented a valid result, the
> `data` entry in the _response map_ should be `null`.

Reviewer: ```suggestion
If an error was raised during execution that propagated to the operation root,
the `data` entry in the _response map_ should be `null`.
```
